Scripts/Python/get_device_list.py

```diff
@@ -34,61 +34,143 @@
-            print("Unable to create a session with appliance %s" % (ip_address))
-    except:
-        print ("Unexpected error:", sys.exc_info()[0])
+            print("Unable to retrieve device list from %s" % self.__session_input["ip"])
+
+        return json_data
+
+    def __get_device_list(self):
+        next_link_url = self.__base_uri + '/api/DeviceService/Devices'
+        self.json_data = None
+
+        while next_link_url is not None:
+            data = self.__get_device_from_uri(next_link_url)
+            next_link_url = None
+            if data['@odata.count'] <= 0:
+                print("No devices managed by %s" % self.__session_input["ip"])
+            if '@odata.nextLink' in data:
+                next_link_url = self.__base_uri + data['@odata.nextLink']
+            if self.json_data is None:
+                self.json_data = data
+            else:
+                self.json_data['value'] += data['value']
+
+    def __format_json(self):
+        # print to console in the absence of a specified file path
+        json_object = json.dumps(self.json_data, indent=4, sort_keys=True)
+        if self.__output_details["path"] == '':
+            print("*** Device List ***")
+            print(json_object)
+            return
+        # If file path is specified then write to file.
+        modified_filepath = self.__get_unique_filename()
+        with open(modified_filepath, "w") as json_file:
+            json_file.write(json_object)
+        json_file.close()
+
+    def __format_csv(self):
+        exportable_props = ["Id", "Identifier", "DeviceServiceTag",
+                            "ChassisServiceTag", "Model", "DeviceName"]
+        generate_default = self.__output_details["path"] == ''
+        if generate_default:
+            self.__output_details["path"] = 'device_list.csv'
+        modified_filepath = self.__get_unique_filename()
+        if generate_default:
+            print("Generating output file({0}) at current location".format(modified_filepath))
+        with open(modified_filepath, 'w', newline='') as csv_file:
+            csv_writer = csv.writer(csv_file)
+            devices = self.json_data["value"]
+            csv_writer.writerow(exportable_props)
+            for device in devices:
+                device_props = []
+                for prop in exportable_props:
+                    device_props.append(device[prop])
+                csv_writer.writerow(device_props)
+        csv_file.close()
+
+    def __get_unique_filename(self):
+        i = 1
+        new_filepath = self.__output_details["path"]
+        exists = os.path.isfile(new_filepath)
+        while os.path.isfile(new_filepath):
+            (root, ext) = os.path.splitext(self.__output_details["path"])
+            new_filepath = root + "({0})".format(i) + ext
+            i += 1
+        if exists:
+            print("Output file exists. Writing to {}".format(new_filepath))
+        return new_filepath
+
+    def __validateargs(self):
```

Comment:
Is this method required? The "required=true" flag for the argument should suffice

For the filename validation you could pass in a method that validates extensions / length of the name etc - see using the "type" example (def perfect_square) - https://docs.python.org/3/library/argparse.html

Reply:
- "required=true" should indeed suffice. Will remove the validation for username and password.

- The "type" parameter can only accept a callable with a single parameter. The filename validation compares the file name against the selected format. Using the "type=callable" approach may not work, since the file format cannot be passed into this callable.